Add the config preset features to the view

seelabs opened the pull request:
When new code is added that is conditional on a feature, that code needs access to the config object. This can mean passing the config object as a parameter through several layers of calls. Adding the preset features to the view solves this problem.

seelabs commented:
Pushed a commit that changes `presets_` from a `shared_ptr` to a ref after feedback from Vinnie. Rational: if someone is using the features after the config object goes away we'd rather know about it (and using a shared_ptr introduces a memory fence as well)

miguelportilla commented on src/ripple/ledger/impl/ReadView.cpp:
```diff
@@ -66,7 +62,7 @@ class Rules::Impl
     enabled (uint256 const& feature) const
     {
-        if (presets_ && presets_->count(feature) > 0)
+        if (presets_.count(feature) > 0)
```
`find` may be a better than `count`.

seelabs replied:
The two functions should be equivalent in terms of performance. I _think_ `count` is a standard idiom for deciding if a set or map contains an element (you see it in our code and in boost::program_options, for example). And `presets_.count(...)` is terser than `presets_.find(..) != presets_.end()`.

Having said all that, this is definitely a nit. I prefer `count`, but I'm happy to change it if you strongly prefer `find`.

ximinez commented on src/ripple/app/misc/impl/TxQ.cpp:
```diff
@@ -1312,8 +1309,7 @@ TxQ::getMetrics(Config const& config, OpenView const& view,
     auto const allowEscalation =
-        (view.rules().enabled(featureFeeEscalation,
-            config.features));
+        (view.rules().enabled(featureFeeEscalation));
```
Some of these changes remove dependencies from `TxQ::getMetrics` and `TxQ::getAccountTxs`. You can cherry pick https://github.com/ximinez/rippled/commit/aecded7487f0159f16cb6d4bf68599c359b73418

seelabs replied:
Awesome! I'll review and cherry pick.

seelabs commented on src/ripple/app/ledger/Ledger.cpp:
```diff
@@ -624,7 +629,7 @@ Ledger::setup (Config const& config)
     try
     {
-        rules_ = Rules(*this);
+        rules_ = Rules(*this, config.features);
```
Yes, I'd like to keep them all. Getting rid of the `rules_(config.features)` in the constructors would require adding a default constructor, and that could cause errors if someone adds code that doesn't add the predefined features. Conversely, the `rules_(config.features)` constructor is not equivalent to the assignment above, so we can't remove that.

But most importantly, even if we could get rid of them, the savings isn't worth it. That isn't _that_ expensive.

ximinez replied:
Excellent point, particularly about the default ctor.

seelabs commented on src/ripple/ledger/impl/ReadView.cpp:
```diff
@@ -81,18 +92,21 @@ class Rules::Impl
+Rules::enabled (uint256 const& id) const
 {
-    if (presets.count(id) > 0)
-        return true;
```
I'm not sure where in the code this comment is referring to. It's marked below some deleted code. Can you clarify where you mean?

ximinez replied:
`Rules::enabled (uint256 const& id) const`

The new first line is `if(!impl_)`. I figure since there's no default ctor now, `impl_` should never be null, so we could `assert(impl_)`.

seelabs replied:
I see. Yes, I agree.

seelabs commented:
Rebased onto 0.60.0-b3 and added two commits.
@miguelportilla Can you review the `Remove Config dependency` commit I cherry picked from Ed? (I reviewed it, thumbs up from me)
@miguelportilla and @ximinez Can you review the last commit `Rules.impl_ is never null`.
Thanks.

miguelportilla commented:
@seelabs On it! I'll review the cherry pick and the last commit today.

miguelportilla commented:
@seelabs The cherry pick from Ed and your last commit look good to me.

seelabs commented:
squashed

ximinez commented:
Merged as https://github.com/ripple/rippled/pull/2005/commits/f5af8b03de8d770d49e444b60ffe2b9063f0ad91
